File: androidApp/src/androidTest/java/com/mbta/tid/mbta_app/android/map/MapboxConfigManagerTests.kt
package com.mbta.tid.mbta_app.android.map

import com.mbta.tid.mbta_app.model.response.ApiResult
import com.mbta.tid.mbta_app.model.response.ConfigResponse
import com.mbta.tid.mbta_app.repositories.MockConfigRepository
import com.mbta.tid.mbta_app.repositories.MockSentryRepository
import com.mbta.tid.mbta_app.usecases.ConfigUseCase
import junit.framework.TestCase.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue
import kotlinx.coroutines.runBlocking
import org.junit.Test

class MapboxConfigManagerTests {

    @Test
    fun testWhenLoadConfigSuccessTokenSet() = runBlocking {
        var configuredToken: String? = null
        val configManager =
            MapboxConfigManager(
                ConfigUseCase(
                    MockConfigRepository(ApiResult.Ok(ConfigResponse("fake_token"))),
                    MockSentryRepository(),
                ),
                configureMapboxToken = { configuredToken = it },
            )

        configManager.loadConfig()

        assertEquals("fake_token", configuredToken)
    }

    @Test
    fun testWhenLoadConfigErrorTokenNotSet() = runBlocking {
        var configureTokenCalled = false
        val configManager =
            MapboxConfigManager(
                configUseCase =
                    ConfigUseCase(
                        MockConfigRepository(ApiResult.Error(500, "oops")),
                        MockSentryRepository(),
                    ),
                configureMapboxToken = { configureTokenCalled = true },
            )

        configManager.loadConfig()

        assertFalse { configureTokenCalled }
    }

    @Test
    fun testInterceptorSetOnInit() {
        var httpInterceptorSet = false
        val configManager =
            MapboxConfigManager(
                configureMapboxToken = {},
                setHttpInterceptor = { httpInterceptorSet = true },
            )
        assertTrue { httpInterceptorSet }
    }
}

File: iosApp/iosAppTests/Utils/MapboxConfigManagerTests.swift
//
//  MapboxConfigManagerTests.swift
//  iosApp
//
//  Created by Brandon Rodriguez on 9/8/25.
//  Copyright © 2025 MBTA. All rights reserved.
//

@testable import iosApp
import Shared
import XCTest

final class MapboxConfigManagerTests: XCTestCase {
    func testWhenLoadConfigSuccessTokenSet() async {
        var configuredToken: String?
        let configRepo = MockConfigRepository(
            response: ApiResultOk(data: ConfigResponse(mapboxPublicToken: "fake_token"))
        )
        let usecase = ConfigUseCase(
            configRepo: configRepo,
            sentryRepo: MockSentryRepository()
        )
        let configManager = MapboxConfigManager(
            configUsecase: usecase,
            configureMapboxToken: { configuredToken = $0 }
        )
        await configManager.loadConfig()
        XCTAssertEqual("fake_token", configuredToken)
    }

    func testWhenLoadConfigErrorTokenNotSet() async {
        var configureTokenCalled = false
        let configRepo = MockConfigRepository(
            response: ApiResultError(code: 500, message: "oops")
        )
        let usecase = ConfigUseCase(
            configRepo: configRepo,
            sentryRepo: MockSentryRepository()
        )
        let configManager = MapboxConfigManager(
            configUsecase: usecase,
            configureMapboxToken: { _ in configureTokenCalled = true }
        )
        await configManager.loadConfig()
        XCTAssertFalse(configureTokenCalled)
    }

    func testInterceptorSetOnInit() {
        var httpInterceptorSet = false
        let configRepo = MockConfigRepository(
            response: ApiResultOk(data: ConfigResponse(mapboxPublicToken: "fake_token"))
        )
        let usecase = ConfigUseCase(
            configRepo: configRepo,
            sentryRepo: MockSentryRepository()
        )
        let configManager = MapboxConfigManager(
            configUsecase: usecase,
            setHttpInterceptor: { _ in httpInterceptorSet = true },
            configureMapboxToken: { _ in }
        )
        XCTAssertTrue(httpInterceptorSet)
    }
}
